src/charm.py

```diff
@@ -1,427 +1,127 @@
-
-    def __init__(self, msg, status_type=StatusBase):
-        super().__init__()
-
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
-
-
-def _gen_obj_storage_endpoint_url(obj_storage):
-    """Generate object storage endpoint URL."""
-    return f"http://{obj_storage['service']}.{obj_storage['namespace']}:{obj_storage['port']}"
-
-
-def _b64_encode_dict(d):
-    """Returns the dict with values being base64 encoded."""
-    # Why do we encode and decode in utf-8 first?
-    return {k: b64encode(v.encode("utf-8")).decode("utf-8") for k, v in d.items()}
-
-
-def _minio_credentials_dict(obj_storage):
-    """Returns a dict of minio credentials with the values base64 encoded."""
-    minio_credentials = {
-        "AWS_ENDPOINT_URL": _gen_obj_storage_endpoint_url(obj_storage),
-        "AWS_ACCESS_KEY_ID": obj_storage["access-key"],
-        "AWS_SECRET_ACCESS_KEY": obj_storage["secret-key"],
-        "USE_SSL": str(obj_storage["secure"]).lower(),
-    }
-    return _b64_encode_dict(minio_credentials)
-
-
-def _seldon_credentials_dict(obj_storage):
-    """Returns a dict of seldon init-container object storage credentials, base64 encoded."""
-    credentials = {
-        "RCLONE_CONFIG_S3_TYPE": "s3",
-        "RCLONE_CONFIG_S3_PROVIDER": "minio",
-        "RCLONE_CONFIG_S3_ACCESS_KEY_ID": obj_storage["access-key"],
-        "RCLONE_CONFIG_S3_SECRET_ACCESS_KEY": obj_storage["secret-key"],
-        "RCLONE_CONFIG_S3_ENDPOINT": _gen_obj_storage_endpoint_url(obj_storage),
-        "RCLONE_CONFIG_S3_ENV_AUTH": "false",
-    }
-    return _b64_encode_dict(credentials)
-
-
-def _db_secret_dict(mysql):
-    """Returns a dict of db-secret credential data, base64 encoded."""
-    db_secret = {
-        "DB_ROOT_PASSWORD": mysql["root_password"],
-        "MLFLOW_TRACKING_URI": f"mysql+pymysql://root:{mysql['root_password']}@{mysql['host']}"
-        f":{mysql['port']}/{mysql['database']}",
-    }
-    return _b64_encode_dict(db_secret)
-
-
-def _get_obj_storage(interfaces):
-    """Unpacks and returns the object-storage relation data.
-
-    Raises CheckFailedError if an anticipated error occurs.
-    """
-    if not ((obj_storage := interfaces["object-storage"]) and obj_storage.get_data()):
-        raise CheckFailedError("Waiting for object-storage relation data", WaitingStatus)
-
-    try:
-        obj_storage = list(obj_storage.get_data().values())[0]
-    except Exception as e:
-        raise CheckFailedError(
-            f"Unexpected error unpacking object storage data - data format not "
-            f"as expected. Caught exception: '{str(e)}'",
-            BlockedStatus,
-        )
-
-    return obj_storage
+            self._check_leader()
+            self._update_layer()
+        except ErrorWithStatus as err:
+            self.model.unit.status = err.status
+            self.logger.info(f"Failed to handle {event} with error: {str(err)}")
```

Comment:
I think this log message is too aggressive.  We're not failing to handle errors here, we get here by *handling* errors.  For example, if this unit is not the leader it is going to say "Failed to handle X with error: 'Not a leader, skipping setup'" and go into `WaitingStatus`.  I like that we log the message, but not that we call it an unhandled error
```suggestion
            self.logger.info(f"Event {event} stopped early with message: {str(err)}")
```